inference-engine/src/multi_device/multi_device_exec_network.cpp

```diff
@@ -60,71 +90,241 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
+        _networkActualNeeded = _acceleratorFuture.get();
+    } else if (_cpuFuture.valid()) {  // only CPU is valid, like AUTO:CPU
+        _networkActualNeeded = _cpuFuture.get();
+    } else {
+        IE_THROW() << "No device task available";
+    }
+
+    // if there is only one device or loading CPU device is failed,
+    // the ActualNetwork is already ok now.
+    if (!_acceleratorFuture.valid()) {
+        _alreadyActualNetwork = true;
+        // setPerfcounts
+        SetPerfCounts(_networkActualNeeded);
+    }
+}
+
+bool MultiDeviceExecutableNetwork::IsActualNetworkReady() const {
+    if (_alreadyActualNetwork) {
+        WaitActualNetworkReady();
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void MultiDeviceExecutableNetwork::WaitActualNetworkReady() const {
+    // Maybe differnt API will call this function, so add call once here
+    // for every MultiDeviceExecutableNetwork instance
+    std::call_once(_oc, [&] () {
+            if (_acceleratorFuture.valid()) {
+                _networkActualNeeded = _acceleratorFuture.get();
+                SetPerfCounts(_networkActualNeeded);
+            }
+            });
+}
+
+void MultiDeviceExecutableNetwork::SetPerfCounts(
+        InferenceEngine::SoExecutableNetworkInternal& executableNetwork) const {
+    try {
+        _needPerfCounters = executableNetwork->GetMetric(PluginConfigParams::KEY_PERF_COUNT).as<std::string>() == PluginConfigParams::YES;
+    } catch (...) {
     }
 }
 
 void MultiDeviceExecutableNetwork::ScheduleToWorkerInferRequest(Task inferPipelineTask, DeviceName preferred_device) {
+    // AUTO work mode
+    if (_workModeIsAUTO) {
+        if (!preferred_device.empty()) {
+            // the preferred_device should be the selected device in AUTO work mode
+            if (preferred_device != _acceleratorDevice.deviceName) {
+                IE_THROW(NotFound) << "The preferred_device should be the selected device";
+            }
+
+            // if the device needed by customer is not ready, need to wait for it
+            WaitActualNetworkReady();
+
+            auto &idleWorkerRequests = _idleWorkerRequests[preferred_device];
+            if (!RunPipelineTask(inferPipelineTask, idleWorkerRequests, preferred_device)) {
+                // no vacant requests this time, storing the task to the respective queue
+                _inferPipelineTasksDeviceSpecific[preferred_device]->push(std::move(inferPipelineTask));
+            }
+        } else {
+            // _acceleratorDevice could be the same as _cpuDevice, such as AUTO:CPU
+            auto &idleWorkerRequests = IsActualNetworkReady()
+                                       ? _idleWorkerRequests[_acceleratorDevice.deviceName]
+                                       : _idleWorkerRequests[_cpuDevice.deviceName];
+            if (!RunPipelineTask(inferPipelineTask, idleWorkerRequests, preferred_device)) {
```

Comment:
the snippet is same as in the upper branch ( when perffered device is not empty).
shall factor this out so that only idleWorkerRequests selection is made within if/else?

Reply:
factor out the RunPipelineTask